aesm-client/src/win_aesm_client.rs

```diff
@@ -0,0 +1,359 @@
+            result: *mut AesmError,
+        ) -> HRESULT,
+    >,
+    get_epid_provision_status: Option<
+        unsafe extern "system" fn(
+            this: *mut AesmInterface,
+            epid_pr_status: *mut u32,
+            result: *mut AesmError,
+        ) -> HRESULT,
+    >,
+    get_platform_service_status: Option<
+        unsafe extern "system" fn(
+            this: *mut AesmInterface,
+            pse_status: *mut u32,
+            result: *mut AesmError,
+        ) -> HRESULT,
+    >,
+}
+
+const CLSID_AESMINTERFACE: CLSID = CLSID {
+    Data1: 0x82367CAB,
+    Data2: 0xF2B9,
+    Data3: 0x461A,
+    Data4: [0xB6, 0xC6, 0x88, 0x9D, 0x13, 0xEF, 0xC6, 0xCA],
+};
+const IID_IAESMINTERFACE: IID = IID {
+    Data1: 0x50AFD900,
+    Data2: 0xF309,
+    Data3: 0x4557,
+    Data4: [0x8F, 0xCB, 0x10, 0xCF, 0xAB, 0x80, 0x2C, 0xDD],
+};
+
+impl Drop for AesmClient {
+    fn drop(&mut self) {
+        unsafe {
+            if let Some(release) = (*(*self.interface).vtbl).release {
+                release(self.interface);
+            }
+            CoUninitialize();
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct AesmClient {
+    interface: *mut AesmInterface,
+}
+
+impl AesmClient {
+    pub fn new() -> Self {
+        AesmClient::create_instance().unwrap()
+    }
+
+    fn create_instance() -> Result<Self> {
+        let mut instance: *mut AesmInterface = std::ptr::null_mut();
+        unsafe {
+            let res = CoInitializeEx(
+                ptr::null_mut(),
+                COINIT_MULTITHREADED | COINIT_DISABLE_OLE1DDE,
+            );
+            if res != S_OK && res != S_FALSE {
+                return Err(Error::AesmCommunication(
+                    IoError::new(ErrorKind::Other,"Fail to initialize Com interface"))
+                );
+            }
+            let res = CoCreateInstance(
+                &CLSID_AESMINTERFACE,
+                ptr::null_mut(),
+                CLSCTX_ALL,
+                &IID_IAESMINTERFACE,
+                &mut instance as *mut _ as *mut *mut c_void,
+            );
+            if res < 0 {
+                return Err(Error::AesmCommunication(
+                    IoError::new(ErrorKind::Other, "Fail to create Aesm Interface"))
+                );
+            }
+        }
+        Ok(AesmClient {
+            interface: instance,
+        })
+    }
+
+    pub fn init_quote(&self) -> Result<QuoteInfo> {
+        let mut target_info: Vec<u8> = vec![0; sgx_isa::Targetinfo::UNPADDED_SIZE];
+        let mut gid: Vec<u8> = vec![0; 4];
+        let mut error: AesmError = 0;
+        unsafe {
+            if let Some(init_quote) = (*(*self.interface).vtbl).init_quote {
+                let ret = init_quote(
+                    self.interface,
+                    target_info.as_mut_ptr(),
+                    target_info.len() as _,
+                    gid.as_mut_ptr(),
+                    gid.len() as _,
+                    &mut error as _,
+                );
```

Comment:
For all calls like these: check `ret` first (and convert to IO error), then check `error`.